core/commonMain/src/implementations/immutableMap/TrieNode.kt

```diff
@@ -876,6 +876,36 @@ internal class TrieNode<K, V>(
         }
     }
 
+    internal fun <K1, V1> equalsWith(that: TrieNode<K1, V1>, equalityComparator: (V, V1) -> Boolean): Boolean {
+        if (this === that) return true
+        if (dataMap != that.dataMap || nodeMap != that.nodeMap) return false
+        if (dataMap == 0 && nodeMap == 0) { // collision node
+            if (buffer.size != that.buffer.size) return false
+            return (0 until buffer.size step ENTRY_SIZE).all { i ->
+                val thatKey = that.keyAtIndex(i)
+                val thatValue = that.valueAtKeyIndex(i)
+                (0 until buffer.size step ENTRY_SIZE).any { j ->
+                    val key = keyAtIndex(j)
+                    val value = valueAtKeyIndex(j)
+                    if (key == thatKey) {
+                        if (equalityComparator(value, thatValue)) true
+                        else return false
```

Comment:
```suggestion
                        return equalityComparator(value, thatValue)
```

Reply:
This is not identical: `return false` here is a shortcut that returns from the whole function, while `true` is just the result of lambda passed to `any` =)

Reply:
Oh, the suggestion is not an equivalent

Reply:
Yep, i'd like to write it in a more obvious and elegant manner, but this code is already kinda' heavy on lambdas within lambdas and I just surrendered to this smelly variant.

Reply:
You can cherry-pick this commit: https://github.com/Kotlin/kotlinx.collections.immutable/commit/72055f7d28b8fe5b915027ac946012d81f93a7e5

Reply:
It seems you still removed the shortcut =)
```kotlin
                val keyIndex = collisionKeyIndex(thatKey)
                if (keyIndex != -1) {
                    val value = valueAtKeyIndex(keyIndex)
                    equalityComparator(value, thatValue)
                } else false
```
should still be
```kotlin
                val keyIndex = collisionKeyIndex(thatKey)
                if (keyIndex != -1) {
                    val value = valueAtKeyIndex(keyIndex)
                    equalityComparator(value, thatValue) || return false
                } else false
```

Reply:
Now that `any { }` function call is removed the `return false` is not a shortcut anymore.
Any `false` result will lead the enclosing `all { }` to return `false` immediately, the result of which is returned by `equalsWith`.